packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java

```diff
@@ -278,6 +301,78 @@ public void onMethodCall(MethodCall call, final Result result) {
           break;
         }
 
+      case "DatabaseReference#runTransaction":
+        {
+          // Tasks are used to allow native execution of doTransaction to wait while Snapshot is
+          // processed by logic on the dart side.
+          TaskCompletionSource<Map<String, Object>> updateMutableDataTCS =
+              new TaskCompletionSource<>();
+          final Task<Map<String, Object>> updateMutableDataTCSTask = updateMutableDataTCS.getTask();
+          final Map<String, Object> arguments = call.arguments();
+
+          // Add transaction task to array for later retrieval.
+          transactionTasks.put(getTransactionKey(arguments), updateMutableDataTCS);
+
+          DatabaseReference reference = getReference(arguments);
+
+          // Initiate native transaction.
+          reference.runTransaction(
+              new Transaction.Handler() {
+                @Override
+                public Transaction.Result doTransaction(MutableData mutableData) {
+                  Map<String, Object> snapshotMap = new HashMap<>();
+                  snapshotMap.put("key", mutableData.getKey());
+                  snapshotMap.put("value", mutableData.getValue());
+
+                  // Return snapshot to dart side for update.
+                  result.success(snapshotMap);
+
+                  try {
+                    // Wait for updated snapshot from the dart side.
+                    Map<String, Object> updatedSnapshotMap = Tasks.await(updateMutableDataTCSTask);
+                    // Set value of MutableData to value returned from the dart side.
+                    mutableData.setValue(updatedSnapshotMap.get("updatedDataSnapshot"));
+                  } catch (ExecutionException | InterruptedException e) {
+                    Log.e(TAG, e.toString());
```

Comment:
It would be useful to abandon the transaction here, this will result in the completion callback having a false `committed` parameter which would indicate that the transaction failed to commit the updates and should be tried again.